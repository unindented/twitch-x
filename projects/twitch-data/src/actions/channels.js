import {getAction} from '../api'

export function loadSearchChannels ({query, limit}) {
  const action = {
    type: 'LOAD_SEARCH_CHANNELS',
    payload: query
  }
  return getAction(action, '/search/channels', {query: {query: encodeURIComponent(query), limit}})
}
