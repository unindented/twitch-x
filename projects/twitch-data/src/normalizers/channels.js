import {normalize, schema} from 'normalizr'
import {valueOrNull} from './utils'

const channel = new schema.Entity('channels', {}, {
  idAttribute: 'name',
  processStrategy: (channel) => {
    return {
      id: valueOrNull(channel.name),
      status: valueOrNull(channel.status),
      name: valueOrNull(channel.display_name),
      game: valueOrNull(channel.game),
      image: valueOrNull(channel.logo)
    }
  }
})

const root = [channel]

export function normalizeSearchChannels (payload) {
  return normalize(payload.channels, root)
}
