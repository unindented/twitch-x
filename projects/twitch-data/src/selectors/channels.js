import {createSelector} from 'reselect'

export const getSearchChannels = createSelector(
  (state) => state.data.channels.search,
  (state) => state.data.channels.byId,
  (searchChannelIds, channelsById) => searchChannelIds.map((id) => channelsById[id])
)
