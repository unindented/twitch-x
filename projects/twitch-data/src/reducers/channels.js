import update from 'immutability-helper'
import {normalizeSearchChannels} from '../normalizers/channels'

const initialState = {
  byId: {},
  search: []
}

export default (state = initialState, action = {}) => {
  switch (action.type) {
    case 'LOAD_SEARCH_CHANNELS_SUCCESS': {
      const {entities, result} = normalizeSearchChannels(action.payload.response)
      return update(state, {$merge: {byId: entities.channels, search: result}})
    }
  }

  return state
}
