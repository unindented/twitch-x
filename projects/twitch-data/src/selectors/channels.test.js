import {getSearchChannels} from './channels'

describe('selectors/channels', () => {
  describe('.getSearchChannels', () => {
    it('returns the search channels when available', () => {
      const state = {
        data: {
          channels: {
            byId: {
              1: {name: 'foo'},
              2: {name: 'bar'},
              3: {name: 'baz'}
            },
            search: [1, 2]
          }
        }
      }
      expect(getSearchChannels(state)).toMatchSnapshot()
    })
  })
})
