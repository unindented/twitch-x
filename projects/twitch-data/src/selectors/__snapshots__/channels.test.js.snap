// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectors/channels .getSearchChannels returns the search channels when available 1`] = `
Array [
  Object {
    "name": "foo",
  },
  Object {
    "name": "bar",
  },
]
`;
